parser-typechecker/src/Unison/Util/Bytes.hs

```diff
@@ -1,42 +1,78 @@
-import System.IO.Unsafe (unsafeDupablePerformIO)
 import Unison.Prelude hiding (ByteString, empty)
-import Basement.Block (Block)
-import qualified Data.ByteString.Lazy as LB
-import qualified Data.ByteArray as B
+import qualified Data.ByteString as B
+import qualified Data.ByteArray as BA
 import qualified Data.ByteArray.Encoding as BE
-import qualified Data.FingerTree as T
+import qualified Data.ByteString.Lazy as LB
 import qualified Data.Text as Text
+import qualified Unison.Util.Rope as R
+import qualified Data.Vector.Primitive as V
+import qualified Data.Vector.Storable as SV
+import qualified Data.Vector.Storable.ByteString as BSV
+import Foreign.Storable (pokeByteOff)
 import qualified Codec.Compression.Zlib as Zlib
 import qualified Codec.Compression.GZip as GZip
 
--- Block is just `newtype Block a = Block ByteArray#`
-type ByteString = Block Word8
+type ByteString = V.Vector Word8
+
+-- Bytes type represented as a rope of ByteStrings
+newtype Bytes = Bytes { underlying :: R.Rope ByteString } deriving (Semigroup,Monoid,Eq,Ord)
 
--- Bytes type represented as a finger tree of ByteStrings.
--- Can be efficiently sliced and indexed, using the byte count
--- annotation at each subtree.
-newtype Bytes = Bytes (T.FingerTree (Sum Int) (View ByteString))
+instance R.Sized ByteString where size = V.length
+instance R.Drop ByteString where drop = V.drop
+instance R.Take ByteString where take = V.take
+instance R.Index ByteString Word8 where index n bs = bs V.!? n
+instance R.Reverse ByteString where reverse = V.reverse
+instance NFData Bytes where rnf _ = ()
 
 null :: Bytes -> Bool
-null (Bytes bs) = T.null bs
+null = R.null . underlying
 
 empty :: Bytes
-empty = Bytes mempty
+empty = mempty
+
+isAscii :: Bytes -> Bool
+isAscii b = all (V.all (<= 0x7F)) (chunks b)
+
+fromByteString :: B.ByteString -> Bytes
+fromByteString b = snoc empty (byteStringToChunk b)
+
+toByteString :: Bytes -> B.ByteString
+toByteString b = B.concat (map chunkToByteString (chunks b))
 
-fromArray :: B.ByteArrayAccess ba => ba -> Bytes
-fromArray = snoc empty
+toArray :: BA.ByteArray b => Bytes -> b
+toArray b = chunkToArray $ V.concat (chunks b)
+
+fromArray :: BA.ByteArrayAccess b => b -> Bytes
+fromArray b = snoc empty (arrayToChunk b)
+
+byteStringToChunk, chunkFromByteString :: B.ByteString -> ByteString
+byteStringToChunk = fromStorable . BSV.byteStringToVector
+chunkFromByteString = byteStringToChunk
+
+chunkToByteString :: ByteString -> B.ByteString
+chunkToByteString = BSV.vectorToByteString . toStorable
+
+fromStorable :: (V.Prim a, SV.Storable a) => SV.Vector a -> V.Vector a
+fromStorable v =
+  -- is there a more efficient implementation of this
+  V.generate (SV.length v) (SV.unsafeIndex v)
+{-# inline fromStorable #-}
+
+toStorable :: (V.Prim a, SV.Storable a) => V.Vector a -> SV.Vector a
+toStorable v =
+  -- is there a more efficient implementation of this
+  SV.generate (V.length v) (V.unsafeIndex v)
+{-# inline toStorable #-}
```

Comment:
@dolio can you help with figuring out the incantation to do one or both of these using a block copy of some sort? 

What's here isn't wrong but I think it can be done using what ends up being a `memcopy`.

Reply:
Done. There are tests for this stuff, right? I had to use some somewhat unsafe stuff.

Reply:
There aren't specific unit tests for these functions but those codepaths should get hit by a bunch of existing tests, yeah.